File: 0001-two-sum/0001-two-sum.py
# return indices of the two numbers such that they add up to target
from collections import defaultdict

class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        dict = defaultdict(list)
        
        for i, n in enumerate(nums):
            dict[n].append(i)
            
        
        for key1 in dict:
            key2 = target - key1
            if key1 == key2 and len(dict[key1]) > 1:
                return [dict[key1][0], dict[key1][1]]
            if key2 in dict and dict[key2] != dict[key1]:
                return [dict[key1][0], dict[key2][0]]

File: 0001-two-sum/0001-two-sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        
        HashMap<Integer, Integer> map = new HashMap<>();
        
        for(int i=0;i<nums.length;i++){
            int num = nums[i];
            
            if(map.get(target-num) != null){
                return new int[]{map.get(target-num), i};
            }else{
                map.put(num, i);
            }
        }
        
        return null;
    }
}